Lib/bisect.py

```diff
@@ -1,25 +1,38 @@
 """Bisection algorithms."""
 
-def insort_right(a, x, lo=0, hi=None):
+def insort_right(a, x, lo=0, hi=None, key=None):
     """Insert item x in list a, and keep it sorted assuming a is sorted.
 
     If x is already in a, insert it to the right of the rightmost x.
 
     Optional args lo (default 0) and hi (default len(a)) bound the
     slice of a to be searched.
+
+    Optional argument key is a function of one argument used to
+    customize the order.
     """
 
+    # the python implementation does not
```

Comment:
The Python implementation does need to be optimized as well (it is used by PyPy, for example).  We usually create two code paths for this:  'if key is None: full logic w/o keyfunc' and 'else: full logic w keyfunc'

Reply:
I'm still looking at this and am considering a different approach of having a separate, fully general function (that includes a *reversed* option) and leaving the core high-performance building blocks as-is and uncluttered.  

We should a usage note advising that the key-function shouldbe wrapped in an lru_cache unless the key-function is already fast.

FWIW, if key-function support is added, it is inevitable that *reversed* support will be requested (since sorted() and bisect() are often used together).